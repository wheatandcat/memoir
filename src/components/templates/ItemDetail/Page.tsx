import React, { memo } from 'react';
import { StyleSheet, ScrollView } from 'react-native';
import View from 'components/atoms/View';
import CardDetail from 'components/organisms/CardDetail/CardDetail';
import InputDateWrap from 'components/organisms/InputDateWrap/InputDateWrap';
import theme from 'config/theme';
import { ItemQuery } from 'queries/api/index';

type Item = NonNullable<ItemQuery['item']>;

type Props = Pick<Item, 'categoryID' | 'title'> & {
  loading: boolean;
  date: string;
  onChangeDate: (date: string) => void;
};

const Page: React.FC<Props> = (props) => {
  return (
    <InputDateWrap
      date={props.date}
      onChangeDate={props.onChangeDate}
      firstItem
    >
      <ScrollView>
        <View style={styles.inner}>
          {props.loading && (
            <CardDetail
              date={props.date}
              title={props.title}
              categoryID={props.categoryID}
            />
          )}
        </View>
      </ScrollView>
    </InputDateWrap>
  );
};

export default memo(Page);

const styles = StyleSheet.create({
  inner: {
    height: '100%',
    marginHorizontal: theme().space(3),
    marginVertical: theme().space(4),
  },
});
