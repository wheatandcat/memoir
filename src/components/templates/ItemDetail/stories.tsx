import React from 'react';
import { storiesOf } from '@storybook/react-native';
import { mockFn } from 'storyBookUtils/index';
import Page from './Page';

storiesOf('templates/ItemDetail', module).add('Page', () => (
  <Page date="2021-02-21" onChangeDate={mockFn} />
));
