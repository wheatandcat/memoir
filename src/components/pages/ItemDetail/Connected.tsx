import React, { memo, useCallback, useState } from 'react';
import { useNavigation } from '@react-navigation/native';
import { useSetRecoilState } from 'recoil';
import dayjs from 'lib/dayjs';
import {
  ItemDocument,
  UpdateItemDocument,
  DeleteItemDocument,
  UpdateItem,
  NewItem,
  DeleteItem,
} from 'queries/api/index';
import { homeDateState } from 'store/atoms';
import useHomeItems from 'hooks/useHomeItems';
import { useQuery, useMutation } from '@apollo/client';
import Plain from './Plain';

export type Props = {
  itemID: string;
  date: string;
};

type State = {
  openUpdateItemModal: boolean;
};

export type ConnectedType = {
  updateItemLoading: boolean;
  date: string;
  openUpdateItemModal: boolean;
  onOpenUpdateItem: () => void;
  onChangeDate: (date: string) => void;
  onUpdateItem: (updateItem: NewItem) => void;
  onDeleteItem: () => void;
  onCloseUpdateItem: () => void;
};

const initialState = (): State => ({
  openUpdateItemModal: false,
});

const Connected: React.FC<Props> = (props) => {
  const navigation = useNavigation();
  const [state, setState] = useState<State>(initialState());
  const setHomeDate = useSetRecoilState(homeDateState);
  const homeItems = useHomeItems();

  const { loading, data, error, refetch } = useQuery(ItemDocument, {
    variables: {
      id: props.itemID,
    },
  });

  const onOpenUpdateItem = useCallback(() => {
    setState((s) => ({ ...s, openUpdateItemModal: true }));
  }, []);

  const onCloseUpdateItem = useCallback(() => {
    setState((s) => ({ ...s, openUpdateItemModal: false }));
  }, []);

  const [updateItemMutation] = useMutation(UpdateItemDocument, {
    async onCompleted() {
      await refetch?.();

      onCloseUpdateItem();
    },
  });

  const onUpdateItem = useCallback(
    (newItem: NewItem) => {
      const updateItem: UpdateItem = {
        id: props.itemID,
        ...newItem,
      };

      const variables = {
        input: updateItem,
      };

      updateItemMutation({ variables });
    },
    [updateItemMutation, props.itemID]
  );

  const [deleteItemMutation] = useMutation(DeleteItemDocument, {
    async onCompleted() {
      await homeItems?.refetch?.();
      navigation.goBack();
    },
  });

  const onDeleteItem = useCallback(() => {
    const deleteItem: DeleteItem = {
      id: props.itemID,
    };
    const variables = {
      input: deleteItem,
    };

    deleteItemMutation({ variables });
  }, [props.itemID, deleteItemMutation]);

  const onChangeDate = useCallback(
    (date: string) => {
      const formatDate = dayjs(props.date).format('YYYY-MM-DD');
      if (formatDate !== date) {
        setHomeDate({
          date: dayjs(date).format('YYYY-MM-DDT00:00:00+09:00'),
        });

        navigation.goBack();
      }
    },
    [props, navigation, setHomeDate]
  );

  return (
    <Plain
      data={data}
      loading={loading}
      error={error}
      updateItemLoading={false}
      date={dayjs(props.date).format('YYYY-MM-DD')}
      openUpdateItemModal={state.openUpdateItemModal}
      onChangeDate={onChangeDate}
      onOpenUpdateItem={onOpenUpdateItem}
      onUpdateItem={onUpdateItem}
      onDeleteItem={onDeleteItem}
      onCloseUpdateItem={onCloseUpdateItem}
    />
  );
};

export default memo(Connected);
