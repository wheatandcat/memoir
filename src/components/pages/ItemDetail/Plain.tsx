import React, { memo } from 'react';
import {
  ItemQuery as Query,
  ItemQueryVariables as Variables,
} from 'queries/api/index';
import ErrorPage from 'components/organisms/Error/Error';
import TemplateItemDetail from 'components/templates/ItemDetail/Page';
import { QueryResult } from '@apollo/client';
import { ConnectedType } from './Connected';

type QueryHookResult = QueryResult<Query, Variables>;
type QueryProps = Pick<QueryHookResult, 'data' | 'loading' | 'error'>;

export type Props = QueryProps & ConnectedType;

const Plain: React.FC<Props> = (props) => {
  if (props.error) return <ErrorPage error={props.error} />;

  const item = props.data?.item;

  return (
    <TemplateItemDetail
      {...props}
      title={item?.title || ''}
      categoryID={item?.categoryID || 0}
      like={item?.like || false}
      dislike={item?.dislike || false}
      itemDate={item?.date || ''}
    />
  );
};

export default memo<Props>(Plain);
