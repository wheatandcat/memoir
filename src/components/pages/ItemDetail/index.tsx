import React, { memo } from 'react';
import { RootStackParamList } from 'lib/navigation';
import { RouteProp } from '@react-navigation/native';
import { StackNavigationProp } from '@react-navigation/stack';
import theme from 'config/theme';
import Connected from './Connected';

type ScreenNavigationProp = StackNavigationProp<
  RootStackParamList,
  'ItemDetail'
>;
type ScreenRouteProp = RouteProp<RootStackParamList, 'ItemDetail'>;

export type Props = {
  navigation: ScreenNavigationProp;
  route: ScreenRouteProp;
};

const ItemDetail: React.FC<Props> = (props) => {
  return (
    <Connected date={props.route.params.date} itemID={props.route.params.id} />
  );
};

export const ItemDetailScreenOption = () => {
  return {
    title: '',
    headerStyle: {
      backgroundColor: theme().color.primary.main,
    },
  };
};

export default memo(ItemDetail);
