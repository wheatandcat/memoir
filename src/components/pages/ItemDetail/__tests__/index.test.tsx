import React from 'react';
import { shallow, ShallowWrapper } from 'enzyme';
import ItemDetail, { Props } from '../';

const propsHomeData = (): Props =>
  ({
    navigation: {
      setParams: jest.fn(),
      navigate: jest.fn(),
    },
    route: {
      params: {
        date: '2020-01-01',
        id: 'test',
      },
    },
  } as any);

describe('components/pages/ItemDetail/index.tsx', () => {
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    wrapper = shallow(<ItemDetail {...propsHomeData()} />);
  });

  it('正常にrenderすること', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
