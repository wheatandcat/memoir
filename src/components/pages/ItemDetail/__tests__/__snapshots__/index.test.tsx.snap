// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/pages/ItemDetail/index.tsx 正常にrenderすること 1`] = `
<Memo(Connected)
  date="2020-01-01"
  itemID="test"
/>
`;
